### PR TITLE
Python 3.9 compatibility

### PR DESCRIPTION
**Pull request by FelixSchwarz**
This is my attempt to provide Python 3.9 compatibility (see #23). The test suite passes and I added a few extra tests which were only present in doctest. This pull request is based on top of PR #24.

The main AST changes are in commit 7f4e56e . 

Please note that I'm totally out of my depth here. I never interacted so closely with Python's AST and  this is mostly based on trial and error.

### CODE REVIEW

**Comment by FelixSchwarz** [comment]
Python 3.9 has been released. Would you mind merging #24 and this PR?

**Comment by cedk** [comment]
LGTM

**Comment by hodgestar** [comment]
> Python 3.9 has been released. Would you mind merging #24 and this PR?

@FelixSchwarz Thank you! I've merged #24. Would you mind merging master into this PR and then I can review it?

**Comment by FelixSchwarz** [comment]
sure, will try to do this later this evening

**Comment by FelixSchwarz** [comment]
I rebased the commits on top of the current master so I hope they are easier to review now :-)

File: genshi/path.py [hunk 1]
```diff
@@ -537,7 +537,7 @@ def __init__(self, text, filename=None, lineno=-1):
                     self.strategies.append(strategy_class(path))
                     break
             else:
-                raise NotImplemented('No strategy found for path')
+                raise NotImplementedError('No strategy found for path')
```
Comment (hodgestar) [inline genshi/path.py:540]: Doh. I wonder how long this silly mistake has been here. Thanks for catching it.

File: genshi/template/astutil.py [hunk 2]
```diff
@@ -82,6 +81,10 @@ def visit(self, node):
             info = True
         except AttributeError:
             info = False
+        if isinstance(node, (str, int)):
```
Comment (hodgestar) [inline genshi/template/astutil.py:84]: I assume many other Python types that appear as constants are possible here? E.g. float, bytes, bool, NoneType, complex.

I guess this counts as a bug in Python 3.9 -- or at least I would complain that a parse tree should consist of AST nodes and not random types. For now could you add the other types I've listed that could appear as constants and possibly file a bug report with Python 3.9 and see what people say?

Comment (hodgestar) [reply genshi/template/astutil.py:84]: Or maybe this is the result of us now calling `visit` on things that were nodes before and now are just constants?

File: genshi/template/astutil.py [hunk 3]
```diff
@@ -729,15 +732,17 @@ def _process_slice(node):
                 if getattr(node, 'step', None):
                     self._write(':')
                     self.visit(node.step)
-            elif isinstance(node, _ast.Index):
+            elif isinstance(node, (_ast.Index, _ast_Constant)):
```
Comment (hodgestar) [inline genshi/template/astutil.py:735]: I am guessing that `_ast_Constant` should have it's only elif that looks like:
```python
elif isinstance(node, _ast_Constant):
    self.visit_Constant(node)
```
However, I'm worried that now almost any kind of AST node could appear here and that instead we need something right at the end that does:
```python
elif isinstance(node, _ast.Node):
    self.visit(node)
```
Here `_ast.Node` is a place holder for a base class for all AST nodes. I'm not 100% such a base class exists.

File: genshi/template/tests/eval.py [hunk 4]
```diff
@@ -480,6 +480,27 @@ def __repr__(self):
             self.assertEqual('index.html', code.co_filename)
             self.assertEqual(50, frame.tb_lineno)
 
+    def test_getitem_with_constant_string(self):
+        data = dict(dict={'some': 'thing'})
+        Expression("dict['some']").evaluate(data)
```
Comment (hodgestar) [inline genshi/template/tests/eval.py:485]: We should assert something about what the result of `.evaluate(data)` returns in these test.

**Review by hodgestar — CHANGES_REQUESTED** [review]
Hello! Thanks again for tackling this. Updating the AST code is a tedious and somewhat complex task.

I think there are only two important issues:

1. With _ast.Index now being gone, the slice handling needs to become more careful and call `.visit(node)` on anything that isn't handled by a special case.

2. The new tests (thank you!) should assert something about the value returned by .evaluate(...).

We should probably also have a few more tests for other new cases (e.g. "data[True]", "data[1+1j]", etc).

Other than those two things, it looks good!

File: genshi/template/astutil.py (hunk 2 above)
Comment (FelixSchwarz) [reply genshi/template/astutil.py:84]: Yes, I suspect we should just not call `visit` on these items but the current code was the easiest thing I could do to fix this. I'll try to check if I can find a more elegant solution.

**Comment by rleigh-codelibre** [comment]
Hi folks, I just wanted to ask if you're considering tagging a 0.7.3 release once this (and any other needed PRs) are merged and Python 3.9 is fully supported?

**Comment by FelixSchwarz** [comment]
Yeah, a new release would be helpful also for me (so Fedora can stop shipping these patches on top of the last release). But we need to fix the AST changes first.

**Comment by hodgestar** [comment]
> Hi folks, I just wanted to ask if you're considering tagging a 0.7.3 release once this (and any other needed PRs) are merged and 
> Python 3.9 is fully supported?

Definitely. Making a new release is not hard these days, so it should be able to happen as soon as this PR lands.

File: genshi/template/tests/eval.py (hunk 4 above)
Comment (FelixSchwarz) [reply genshi/template/tests/eval.py:485]: You mean something like 632a0483c49ff9547960324c09ff6093d2ab91de ?

Comment (hodgestar) [reply genshi/template/tests/eval.py:485]: Yep! Perfect.

**Comment by FelixSchwarz** [comment]
I think I didn't mention the relevant pieces from the [Python 3.9 Release Notes](https://docs.python.org/3/whatsnew/3.9.html) regarding the AST changes:

> Simplified AST for subscription. Simple indices will be represented by their value, extended slices will be represented as tuples. `Index(value)` will return a `value` itself, `ExtSlice(slices)` will return `Tuple(slices, Load())`. (Contributed by Serhiy Storchaka in [bpo-34822](https://bugs.python.org/issue34822).)

> ast classes `slice`, `Index` and `ExtSlice` are considered deprecated and will be removed in future Python versions. `value` itself should be used instead of `Index(value)`. `Tuple(slices, Load())` should be used instead of `ExtSlice(slices)`. (Contributed by Serhiy Storchaka in [bpo-34822](https://bugs.python.org/issue34822).)

I plan on extending the commit message once we have a nice changeset as I think this might provide some valuable context for the commit contents. Again - feel free to improve the code as I don't have any experience working with the AST (and the little bit I gained vanished in the months since I first created this PR).

File: genshi/template/astutil.py (hunk 3 above)
Comment (FelixSchwarz) [reply genshi/template/astutil.py:735]: Good catch about `self.visit_Constant(node)`, will change that for sure.

**Comment by FelixSchwarz** [comment]
I added a few commits but it still feels a bit "clumsy" (maybe a more able person might understand the actual meaning of Python's AST changes better). Feedback welcome.

**Review by hodgestar — APPROVED** [review]
I think this is good for now. Thanks again for dealing with the AST changes!

**Pull request merged**